crates/anvil-polkadot/src/substrate_node/service/mod.rs

```diff
@@ -48,11 +69,228 @@ pub struct Service {
+) -> impl Fn(
+    Hash,
+    (),
+) -> futures::future::Ready<
+    Result<
+        (sp_timestamp::InherentDataProvider, MockValidationDataInherentDataProvider<()>),
+        Box<dyn std::error::Error + Send + Sync>,
+    >,
+> + Send
++ Sync {
+    move |block: Hash, ()| {
+        let current_para_head = client
+            .header(block)
+            .expect("Header lookup should succeed")
+            .expect("Header passed in as parent should be present in backend.");
+
+        let slot_duration = client.runtime_api().slot_duration(current_para_head.hash()).unwrap();
+
+        let para_id = client.runtime_api().parachain_id(current_para_head.hash()).unwrap();
+
+        let current_para_block_head =
+            Some(polkadot_primitives::HeadData(current_para_head.encode()));
+
+        let next_block_number =
+            UniqueSaturatedInto::<u32>::unique_saturated_into(current_para_head.number) + 1;
+
+        let next_time = time_manager.current_call_timestamp().checked_mul(1000).unwrap();
+
+        let slot = next_time.saturating_div(slot_duration.as_millis());
+
+        let additional_key_values =
+            vec![(relay_chain::well_known_keys::CURRENT_SLOT.to_vec(), Slot::from(slot).encode())];
+
+        let mocked_parachain = MockValidationDataInherentDataProvider::<()> {
+            current_para_block: next_block_number,
+            para_id,
+            current_para_block_head,
+            relay_offset: next_time as u32,
+            relay_blocks_per_para_block: 1,
+            para_blocks_per_relay_epoch: 10,
+            additional_key_values: Some(additional_key_values),
+            ..Default::default()
+        };
+
+        let timestamp_provider = sp_timestamp::InherentDataProvider::new(
+            next_time.into(),
+        );
+
+        futures::future::ready(Ok((timestamp_provider, mocked_parachain)))
+    }
+}
+
 /// Builds a new service for a full client.
 pub fn new(
     anvil_config: &AnvilNodeConfig,
-    config: Configuration,
+    mut config: Configuration,
 ) -> Result<(Service, TaskManager), ServiceError> {
+    if let Some(ref fork_url) = anvil_config.eth_rpc_url {
+        let http_url = fork_url.clone();
+        let spec_or_top =
+            std::thread::spawn(move || -> eyre::Result<Result<Vec<u8>, Map<String, Value>>> {
+                let rt = TokioRtBuilder::new_current_thread()
+                    .enable_all()
+                    .build()
+                    .map_err(|e| eyre::eyre!("tokio rt build error: {e}"))?;
+                rt.block_on(async move {
+                    let mut headers = HeaderMap::new();
+                    headers
+                        .insert("Accept-Encoding", HeaderValue::from_static("gzip, deflate, br"));
+                    let http = HttpClientBuilder::default()
+                        .set_headers(headers)
+                        .build(http_url)
+                        .map_err(|e| eyre::eyre!("http client build error: {e}"))?;
```

Comment:
I would replace this with subxt::RpcClient. Feels a bit more compact: https://docs.rs/subxt/latest/subxt/backend/rpc/struct.RpcClient.html.

Reply:
I actually thought we can also keep it around. The idea is simple - for fetching the entire state from an existing local chain. Would lazy loading backend allow it? This is only for the case where such thing would be useful of course, and since we have some code for it, it would be easy to just replace the lazy loading with a "sync" loading, when needed. But I am ok with also dropping the code.